Add IMEX ImplicitDenseOutput mutator

wthrowe opened the pull request:
## Proposed changes

<!--
At a high level, describe what this PR does.
-->

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

wthrowe committed:
Add IMEX ImplicitDenseOutput mutator

wthrowe committed:
Add History complete_step_start method

Attempt to retrieve the step start in a time-stepper-agnostic way.
Should work as long as we don't add single-stage methods that try to
use FSAL or something.

wthrowe committed:
Change dense_update_u to add instead of overwrite

IMEX needs to update from multiple histories for a single output.

wthrowe commented:
Please hold off on merging.  There is a bug somewhere in the IMEX dense output code that needs tracking down.  I don't know if it in this PR or somewhere else.

nilsdeppe reviewed (changes requested):
Waiting on Will. This way it can't get accidentally merged :)

wthrowe commented:
I haven't figured out what's going on yet, but I'm pretty much convinced that it's not related to this code, so I think we can merge this.  I've added a check that the dense output returned success and rebased on develop.
